clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -639,19 +647,31 @@ private void maybeUpdate(long now, Node node) {
                 ClientRequest clientRequest = request(now, nodeConnectionId, metadataRequest);
                 log.debug("Sending metadata request {} to node {}", metadataRequest, node.id());
                 doSend(clientRequest, now);
-            } else if (connectionStates.canConnect(nodeConnectionId, now)) {
+                return requestTimeoutMs;
+            }
+
+            if (lastConnectAttemptNodeId != null &&
```

Comment:
Thanks for the explanation @kawamuray. The approach makes more sense to me now. 

I think the scenario we're talking about is when we attempt to connect to a broker and it fails immediately. Depending on how we interpret `reconnect.backoff.ms`, we could either delay reconnecting to just the one node or to any node. The latter seems to have a simpler implementation, but the former is probably the behavior users would expect. So I'm wondering if we could keep the current approach, but move the complexity. For example, instead of keeping track of the node we last tried to connect to here, what if we exposed a method in `NetworkClient` which let tells us if there are _any_ nodes that are still connecting. If there are, then we don't initiate any new connections. So basically the logic would be something like this:

1. Check for any available nodes that we can use now.
2. If there are none, check if there are any connections underway.
3. If there are none, initiate a new connection.

Do you think that would simplify the logic at all?